### ISSUE TITLE
Possibility of "null" version IDs on buckets with versioning

### ISSUE THREAD

**Issue by rluetzner** [minio/minio#14991]
## Expected Behavior

Versioned buckets, especially in case of object locking, should ensure that all objects are created with a valid version.

## Current Behavior

In very rare cases objects in versioned buckets are created with a "null" version. Even with object locking enabled on the bucket, these versions can be deleted. No deletion marker will be left, instead the entire data is gone.

This happens consistently for one of our customers. We've forked and modified the code and are using GlusterFS as a backend storage (see https://github.com/iternity-dotcom/minio/issues/24).

**Nevertheless** I'm pretty sure that this might affect the official MinIO code as well. Look at this excerpt from `bucket-versioning.go`:

```golang
// Enabled enabled versioning?
func (sys *BucketVersioningSys) Enabled(bucket string) bool {
	vc, err := globalBucketMetadataSys.GetVersioningConfig(bucket)
	if err != nil {
		return false
	}
	return vc.Enabled()
}
```

Any error while trying to determine if the bucket is versioned will cause the server to think that versioning is disabled. I cannot exactly describe the root cause that causes the "null" versions for our customers, but I think that there's a transient error that prevents the bucket config from being read from `.minio.sys/buckets/<bucket>/.metadata.bin`. An error with the backend should otherwise prevent the object from being created, but if it is transient enough (and at the exact right moment) the bucket will _disable_ versioning and allow the object to be created.

## Possible Solution

I think the best course of action is to log and/or return the error. As far as I could see the bucket config is always created. It might be missing in legacy cases, but then the error handling should be more specific. If the config is inaccessible and we're running an erasure coded setup there's no way to be sure that we're not working with a versioned bucket.

To ensure that objects are versioned properly I suggest that the request should fail and be retried on the client side in case of such an error.

## Steps to Reproduce (for bugs)

See my theory above. I have no clear way of reproducing this error. Sorry.

## Context

Objects in object locked buckets are created with "null" version IDs and can be permanently deleted, even if a global retention is configured.

## Regression

No. I think this bug was hidden all along.

## Your Environment

The code still exists on the current `master`.

**Comment by harshavardhana** [comment]
> This happens consistently for one of our customers. We've forked and modified the code and are using GlusterFS as a backend storage (see [iternity-dotcom#24](https://github.com/iternity-dotcom/minio/issues/24)).

@rluetzner you should never be using GlusterFS as a backend for MinIO erasure-coded setups.  That is the biggest mistake, you need a strictly consistent backend not a networked filesystem such as GlusterFS. 

As long as the metadata exists on the disk and can be returned for a bucket is under versioning there is no reason it will create a `null` version. 

NOTE: Let's say if we can't read that object, we can't possibly imagine things at this point either since it's a single metadata.bin file - if it is not readable by some means we do allow access to such a bucket anyways.

However, retention should still be applied on `null` versions even if this happens, which we should fix.

**Comment by klauspost** [comment]
@rluetzner You do realize that by offering an AGPL licensed product your customers are subject to AGPL license restrictions as well?

**Comment by rluetzner** [comment]
@klauspost AFAIK there's no license violation there. The complete source code and all changes we made are publicly available under the same license.

@harshavardhana I'm still convinced that the code I posted above should at least _log_ an error and not silently swallow it. It allows for unintended behavior and is hard to track.

**Comment by klauspost** [comment]
@rluetzner AGPL extends to any use of the service, not just the service itself. I suggest you have your legal team examine this.

**Comment by harshavardhana** [comment]
> @harshavardhana I'm still convinced that the code I posted above should at least _log_ an error and not silently swallow it. It allows for unintended behavior and is hard to track.

Yeah, logging an error is okay, but it doesn't help anything here - I suspect that you are using GlusterFS here is the problem. What is happening here is that we expect the written data to be synchronous O_SYNC and it should never return any state information.  If it does then it's a bigger issue. 

We haven't seen this with any of our regular distributed deployments. So I suspect there is something wrong with the fork here.

**Comment by rluetzner** [comment]
This issue is not about our specific setup. That's just what brought me to this piece of code. I'm not trying to leverage you to solve any issues with our fork. I'm trying to help improve your code.

If you're sure that this can never happen then the error handling code can just as well panic or return an error or just not be there at all. Otherwise, if it *does* happen for whatever reason (maybe a regression) you'll have a hard time trying to figure out what happenend. IMHO it's better to panic and have the client retry rather than continuing on possibly false information.

If there is a specific error that you're expecting that specifically means versioning is disabled, then you should handle that error explicitly.

**Comment by harshavardhana** [comment]
Correct @rluetzner the problem is we can't imagine if a bucket has versioning or not if the information is stale. We can't know if the information is stale or not. 

If anything that was persisted on the disk which is when you wanted to enable versioning or retention - it would have succeeded.  If it did indeed succeed then there is no way that any other server will have stale information here.

The issue is something that was written was wrong, or the underlying backend is returning stale information in both scenarios it is the I/O path that is wrong here and should be fixed.

However we will verify why `null` versions don't get retention applied - since it should regardless of the `versionId`.

**Comment by rluetzner** [comment]
Sorry, but I can't follow you. To me it still feels like you're trying to analyze my storage problem. 😅 I'm very sorry for causing so much confusion.

I came here to discuss a possible bug in the code excerpt I posted above. I fully agree with you, that in case of an error the underlying storage is the issue and needs to be fixed. I do also understand that such a bug is very unlikely to happen in any use cases you support *right now*.

All I'm trying to say is that this error can easily go unnoticed because you're handling the error in a dangerous way. Even if you start logging the error, this function is called very early from inside PutObject. It would make sense to safely continue if the object had already been persisted, but I find it makes little sense to continue otherwise, because the bucket config should always exist and be readable. Anything else is a bug and that should at least be easy to detect. I'm struggling to understand why this function has to return a default value and cannot be allowed to fail.

----

FYI (and besides the point), the versioning info can never be stale on object locked buckets, because versioning cannot be disabled. 😉

**Comment by harshavardhana** [comment]
> Sorry, but I can't follow you. To me it still feels like you're trying to analyze my storage problem. 😅 I'm very sorry for causing so much confusion.
> 
> I came here to discuss a possible bug in the code excerpt I posted above. I fully agree with you, that in case of an error the underlying storage is the issue and needs to be fixed. I do also understand that such a bug is very unlikely to happen in any use cases you support _right now_.
> 
> All I'm trying to say is that this error can easily go unnoticed because you're handling the error in a dangerous way. Even if you start logging the error, this function is called very early from inside PutObject. It would make sense to safely continue if the object had already been persisted, but I find it makes little sense to continue otherwise, because the bucket config should always exist and be readable. Anything else is a bug and that should at least be easy to detect. I'm struggling to understand why this function has to return a default value and cannot be allowed to fail.
> 
> FYI (and besides the point), the versioning info can never be stale on object locked buckets, because versioning cannot be disabled. 😉

Yes agreed let me stare at it more closely, see if we can definitively never return false unless the error was indeed that versioning was never set. 

Currently it's not returning because of the limited errors it would generate. A panic() is perhaps a good idea..

**Comment by harshavardhana** [comment]
This should be sufficiently addressed with #14993 and #14994
